### ISSUE TITLE
ROS2 port

### ISSUE THREAD

**Issue by jonbinney** [ros-perception/laser_assembler#17]
Creating this issue as a placeholder for discussions about design decisions when porting to ROS2

**Comment by vandanamandlik** [comment]
Hi,
I couldn't find point_cloud_conversion.h file in ROS2 sensor_msgs package.
Does anyone know any alternative for this file ?

**Comment by jonbinney** [comment]
Not sure - if you haven't already, I'd suggest adding an issue on the ROS2 sensor_msgs package, since that is where it would go.

**Comment by vandanamandlik** [comment]
Hi,
In current laser assembler package there is support for both PointCloud and PointCloud2. So Should I keep it as it is or remove PointCloud support ?

**Comment by jonbinney** [comment]
It would be good to keep support for both. It looks like ROS2 still has a PointCloud message: https://github.com/ros2/common_interfaces/tree/master/sensor_msgs/msg

The PointCloud message is simpler to use for people who aren't using PCL, since the PointCloud2 is a big binary blob that you have to manually parse.

**Comment by vandanamandlik** [comment]
As laser assembler package is dependent on filters package, We are also working on filters porting.
So should we create separate discourse thread for filters or comment under laser assembler thread itself ?

**Comment by jonbinney** [comment]
That should be a separate thread - filters is used in lots of places other than laser_assembler.

**Comment by vandanamandlik** [comment]
Created separate thread for filters package here https://discourse.ros.org/t/filters-porting-to-ros2/7091

**Comment by tonaaskarwur** [comment]
Is there any current updates about the ROS2 version of this pkg?

**Comment by jonbinney** [comment]
This fell off of my radar. I'll take a look later this week.

**Comment by Marina-Banov** [comment]
Any updates for ROS2?

**Comment by lukechencqu** [comment]
Currently, only laser_geometry has ros2 version, but when will this laser_assembler will be available for ROS2?

**Comment by jonbinney** [comment]
Sorry I haven't had time to work on a ROS2 port. If anyone else wants to tackle it, I'm happy to review a PR.

**Comment by JSOV2001** [comment]
Hi,

I'm using ROS2 Humble for a project. I've been trying to use ros2 branch but It's no working. Can someone help me, please?

**This is the error:**

CMake Deprecation Warning at CMakeLists.txt:1 (cmake_minimum_required):
  Compatibility with CMake < 2.8.12 will be removed from a future version of
  CMake.

  Update the VERSION argument <min> value or use a ...<max> suffix to tell

  CMake that the project does not need compatibility with older versions.

CMake Error at CMakeLists.txt:11 (find_package):

  By not providing "Findcatkin.cmake" in CMAKE_MODULE_PATH this project has

  asked CMake to find a package configuration file provided by "catkin", but

  CMake did not find one.

  Could not find a package configuration file provided by "catkin" with any
  of the following names:

    catkinConfig.cmake

    catkin-config.cmake

  Add the installation prefix of "catkin" to CMAKE_PREFIX_PATH or set
  "catkin_DIR" to a directory containing one of the above files.  If "catkin"
  provides a separate development package or SDK, be sure it has been
  installed.

---

Failed   <<< laser_assembler [0.74s, exited with code 1]

Summary: 0 packages finished [1.04s]

  1 package failed: laser_assembler

  1 package had stderr output: laser_assembler

  1 package not processed

**Comment by aarjan222** [comment]
hello @JSOV2001 was this problem of yours solved or not because I also needed this package for my project and when I was building with colcon. It failed saying same error as of yours. It might be because the package is instead searching for catkin to build the package. 

Had you solved this problem at that time? If you found a solution back then, could you please provide some hints or assistance?  

I would be very grateful for your help.
Thank you!

**Comment by jonbinney** [comment]
Yeah, this still isn't ported to ROS2. Could you each explain your use case so I have a sense for what functionality is most critical, and what might already be covered by other packages?

**Comment by JSOV2001** [comment]
> hello @JSOV2001 was this problem of yours solved or not because I also needed this package for my project and when I was building with colcon. It failed saying same error as of yours. It might be because the package is instead searching for catkin to build the package.
> 
> Had you solved this problem at that time? If you found a solution back then, could you please provide some hints or assistance?
> 
> I would be very grateful for your help. Thank you!

Hi, @aarjan222 

First, I'll make clarification: I had to convert my LaserScan message to PointCloud2 message beforehand. Also, in my case, I used the _library PCL_ to do the PointCloud2 processing. Please, remember that PCL is written in C++ code.

So:

- I took the ROS' PointCloud2 message. 
- Then, converted into a PCL's PointCloud2 data. 
- Finally I processed the data in the way my application needed.

In my case, I concatenated every current PointCloud2 into a empty PointCloud2, and then, I got the final result.

Link por PCL: https://pointclouds.org/

**Comment by jonbinney** [comment]
I see - that makes sense. There's some overlap between a few different packages:
- commoon_interfaces/sensor_msgs : has an [iterator](https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/include/sensor_msgs/point_cloud2_iterator.h) for simple access and modification of sensor_msgs without needing a PCL dependency
- perception_pcl/pcl_conversions:[ contains methods ](https://github.com/ros-perception/perception_pcl/blob/ros2/pcl_conversions/include/pcl_conversions/pcl_conversions.h) for converting between sensor_msgs/Pointcloud2 message and pcl Pointcloud type

@JSOV2001  Is i it specifically the sensor_msgs/Laserscan to pcl Pointcloud conversion method that you're missing?

**Comment by JSOV2001** [comment]
> I see - that makes sense. There's some overlap between a few different packages:
> - commoon_interfaces/sensor_msgs : has an [iterator](https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/include/sensor_msgs/point_cloud2_iterator.h) for simple access and modification of sensor_msgs without needing a PCL dependency
> - perception_pcl/pcl_conversions:[ contains methods ](https://github.com/ros-perception/perception_pcl/blob/ros2/pcl_conversions/include/pcl_conversions/pcl_conversions.h) for converting between sensor_msgs/Pointcloud2 message and pcl Pointcloud type
> 
> @JSOV2001  Is i it specifically the sensor_msgs/Laserscan to pcl Pointcloud conversion method that you're missing? 

Hi, Jon

In my case, I transformed the ROS' LaserScan into ROS' PointCloud2 with a function of my own

I base my idea from one of the laser_assembler's package, especifically **ConvertToCloud**

**Comment by jonbinney** [comment]
@JSOV2001 your use case does seem to be missing from the packages I mentioned, and it is an important use case.  We could add the LaserScan->Pointcloud2 functionality to another package:
- adding it to pcl_conversions would require users to pull in PCL as a dependency, which it sounds like you do anyway, but others might not. 
- adding it to sensor_msgs would require that package to depend on laser_geometry, as well as filters if we include the filter plugin aspect
- adding it to laser_filters could work..... it wouldn't add many new dependencies.

Thoughts?

**Comment by jonbinney** [comment]
(i'm trying to avoid adding an extra package in ROS2 unless it really is necessary)

**Comment by ucesiox** [comment]
Hi there, 

I'm just looking for a way to ring buffer point cloud messages of a single livox mid360 so I can get aggregated point clouds over the last n seconds output onto another topic. Don't need any filtering transforming or anything else. 

This repo looked like it would have done the job at some point, but wondering what the current state is! Or whether you know of any other packages which might do the job! Or whether I should just do it myself :P  

Thanks!

**Comment by jonbinney** [comment]
@ucesiox i think for now you're best off implementing it yourself, sorry!

**Comment by Danendra10** [comment]
Is the porting to ros2 not available yet? I was wondering about contributing to port the version on humble through PR, do you have any standarization according the code formats, tests, etc?

**Comment by jonbinney** [comment]
@Danendra10 what's your use case? If you need to go from laserscan to pointcloud2, does the L[aserProjection class in the laser_geometry package](https://github.com/ros-perception/laser_geometry/blob/rolling/include/laser_geometry/laser_geometry.hpp) may cover what you need, and it is already released in ros2 i believe.

**Comment by Danendra10** [comment]
@jonbinney In my case, I need to assembly pointcloud2 from lidar to a pointcloud2, and also there would be multiple lidar. And as far as I know, the laser geometry only have the sensor_msgs/LaserScan type isn't it?

**Comment by jonbinney** [comment]
Yes, I think you're right about geometry_msgs only handling 2d laser scans. I don't have much experience with 3D lidars, but I think they generally publish Pointclouds directly. You could look at the existing 3d lidar drivers (velodyne, for instance) to see how they convert from raw data from the sensor to a PointCloud2. In the velodyne case, I think it directly outputs X,Y,Z points, so it is just assembling those into one big point cloud.

Do you want to publish data from multiple lidars in one PointCloud2 message? That would lose some information, since the lidar is telling you not just that there is something at the end of each beam, but also that there is free space between the origin of the scanner and the end of each beam. If you have multiple scanners, only one of them could be at the end of the beam. I'd suggest publishing one Pointcloud2 for each lidar.

It may be easiest for you to look at the PCL library for helper functions to create the pointcloud2 from altitude/azimuth/range info ( i don't have experience with PCL lately but it seems like something they would have).
